### PR TITLE
[Bugfix] Fix super method lookup for unbounded added methods

### PR DESCRIPTION
**Pull request by itarato**
Fixes: https://github.com/oracle/truffleruby/issues/3131.

### CODE REVIEW

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java [hunk 1]
```diff
@@ -1515,7 +1515,8 @@ private RubySymbol addInternalMethod(RubyModule module, String name, InternalMet
 
             final Visibility visibility = DeclarationContext
                     .findVisibilityCheckSelfAndDefaultDefinee(module, callerFrame);
-            module.addMethodConsiderNameVisibility(getContext(), method, visibility, this);
+            module.addMethodConsiderNameVisibility(getContext(), method.withOwner(module).withDeclaringModule(module),
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleNodes.java:1518]: Since we change `owner` here we should ensure that is the correct semantics, and and so also test `.owner` in the new spec.

File: spec/ruby/core/module/define_method_spec.rb [hunk 2]
```diff
@@ -133,6 +133,25 @@ class << klass
...
+        def bar
+          "super_" + super
+        end
+      end
+
+      foo = foo_class.new
+      foo.singleton_class.send(:define_method, :bar, mixin_module.instance_method(:bar))
+
+      foo.bar.should == "super_bar"
+    end
```
Comment (andrykonchin) [inline spec/ruby/core/module/define_method_spec.rb:154]: **minor:** it seems a `spec/ruby/language/super_spec.rb` file is a better place for such test.

Comment (itarato) [reply spec/ruby/core/module/define_method_spec.rb:154]: What would be the rational for that? Checking `super` here is just a way to prove that `define_method` works as expected, hence it's secondary to the test.

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java (hunk 1 above)
Comment (itarato) [reply src/main/java/org/truffleruby/core/module/ModuleNodes.java:1518]: Sure thing.

File: spec/ruby/core/module/define_method_spec.rb (hunk 2 above)
Comment (andrykonchin) [reply spec/ruby/core/module/define_method_spec.rb:155]: There are a lot of tests for corner cases tested in this particular file. There is a case for `define_method` as well.

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
Looks great, thanks. I'll let @andrykonchin integrate this one once the small things I commented on are addressed.

File: spec/ruby/core/module/define_method_spec.rb [hunk 3]
```diff
@@ -133,6 +133,17 @@ class << klass
...
       end
     end
+
+    it "is sets the method owner for a dynamically added method with a different original owner" do
+      mixin_module = Module.new do
+        def bar; end
+      end
+
+      foo = Object.new
+      foo.singleton_class.send(:define_method, :bar, mixin_module.instance_method(:bar))
```
[the remark this answers is not in the document]
Comment (itarato) [reply spec/ruby/core/module/define_method_spec.rb:143]: Updated

File: spec/ruby/core/module/define_method_spec.rb [hunk 4]
```diff
@@ -133,6 +133,17 @@ class << klass
         klass.should have_public_instance_method(:baz)
       end
     end
+
+    it "is sets the method owner for a dynamically added method with a different original owner" do
```
[the remark this answers is not in the document]
Comment (itarato) [reply spec/ruby/core/module/define_method_spec.rb:137]: Updated

File: spec/ruby/language/super_spec.rb [hunk 5]
```diff
@@ -203,6 +203,25 @@ def a(arg)
...
+    end
+
+    mixin_module = Module.new do
+      def bar
+        "super_" + super
+      end
+    end
+
+    foo = foo_class.new
+    foo.singleton_class.send(:define_method, :bar, mixin_module.instance_method(:bar))
```
[the remark this answers is not in the document]
Comment (itarato) [reply spec/ruby/language/super_spec.rb:220]: Updated

**Comment by itarato** [comment]
Failing CI error will be fixed by a separate PR: https://github.com/oracle/truffleruby/pull/3107

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
> Failing CI error will be fixed by a separate PR: #3107

Yeah, that's separate and already on master, no need to wait for that to merge this.
